src/libcore/iter/range.rs

```diff
@@ -5,175 +5,426 @@ use crate::usize;
+    /// * `Step::backward(a, n) == (0..n).fold(a, |x, _| Step::backward(x, 1))`
+    ///   * Corollary: `Step::backward(a, 0) == a`
+    /// * `Step::backward(a, n) <= a`
+    /// * `Step::forward(Step::backward(a, n), n) == a`
+    #[unstable(feature = "step_trait_ext", reason = "recently added", issue = "42168")]
+    fn backward(start: Self, count: usize) -> Self {
+        Step::backward_checked(start, count).expect("overflow in `Step::backward`")
+    }
 
-    /// Subtracts a `usize`, returning `None` on underflow.
-    fn sub_usize(&self, n: usize) -> Option<Self> {
-        // this default implementation makes the addition of `sub_usize` a non-breaking change
-        let _ = n;
-        unimplemented!()
+    /// Returns the value that would be obtained by taking the *predecessor*
+    /// of `self` `count` times.
+    ///
+    /// # Safety
+    ///
+    /// It is undefined behavior for this operation to overflow the
+    /// range of values supported by `Self`. If you cannot guarantee that this
+    /// will not overflow, use `backward` or `backward_checked` instead.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`:
+    ///
+    /// * if there exists `b` such that `b < a`, it is safe to call `Step::backward_unchecked(a, 1)`
+    /// * if there exists `b`, `n` such that `steps_between(&b, &a) == Some(n)`,
+    ///   it is safe to call `Step::backward_unchecked(a, m)` for any `m <= n`.
+    ///
+    /// For any `a` and `n`, where no overflow occurs:
+    ///
+    /// * `Step::backward_unchecked(a, n)` is equivalent to `Step::backward(a, n)`
+    #[unstable(feature = "unchecked_math", reason = "niche optimization path", issue = "none")]
+    unsafe fn backward_unchecked(start: Self, count: usize) -> Self {
+        Step::backward(start, count)
     }
 }
 
 // These are still macro-generated because the integer literals resolve to different types.
 macro_rules! step_identical_methods {
     () => {
         #[inline]
-        fn replace_one(&mut self) -> Self {
-            mem::replace(self, 1)
+        unsafe fn forward_unchecked(start: Self, n: usize) -> Self {
+            start.unchecked_add(n as Self)
         }
 
         #[inline]
-        fn replace_zero(&mut self) -> Self {
-            mem::replace(self, 0)
+        unsafe fn backward_unchecked(start: Self, n: usize) -> Self {
+            start.unchecked_sub(n as Self)
         }
+    };
+    ( [$u:ident $i:ident] ) => {
+        step_identical_methods!();
 
         #[inline]
-        fn add_one(&self) -> Self {
-            Add::add(*self, 1)
+        fn forward(start: Self, n: usize) -> Self {
+            match Self::forward_checked(start, n) {
+                Some(result) => result,
+                None => {
+                    let result = Add::add(start, n as Self);
+                    // add one modular cycle to ensure overflow occurs
+                    Add::add(Add::add(result as $u, $u::MAX), 1) as Self
+                }
+            }
```

Comment:
Yes, this single modular cycle gets optimized out completely: https://rust.godbolt.org/z/85F3jc

I've added a comment stating this.

Reply:
Ah, wait, you're talking about the whole method, of course. I think I checked this previously, but checking again...

Reply:
[No, it seems that this formulation doesn't optimize down to just an add.](https://rust.godbolt.org/z/y_f2dR)

Reply:
[This spelling that doesn't reuse `forward_checked` does optimize to a single add.](https://rust.godbolt.org/z/QV3HVR) However, I'm less certain that it is correct for every data type (including those bigger than `usize`).

```rust
pub fn forward(start: u16, n: usize) -> u16 {
    if n > u16::MAX as usize || n as u16 > u16::MAX {
        let result = Add::add(start, n as u16);
        // add one modular cycle to ensure overflow happens
        Add::add(Add::add(result as u16, u16::MAX), 1) as u16
    } else {
        Add::add(start, n as u16)
    }
}
```

Reply:
[This definition here is definitely wrong for signed types.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8632161fb631d26385dc1308d29ff6ba)

I'm open to someone figuring out a formulation that both optimizes to just a single add and is correct for all numeric primitive types, but I can't think of one.

Reply:
Here's my attempt:
```rust
pub fn forward(start: i16, n: usize) -> i16 {
    // For debug builds only, trigger an overflow panic
    if forward_checked(start, n).is_none() {
        Add::add(u16::MAX, u16::MAX);
    }
    Add::add(start, n as i16)
}
```

Reply:
Do these methods even get recompiled in the stdlib consumer context?

This definition does seem to work for all cases, and optimizes better than the previous, so I added it. It would probably make sense to retarget the rust-timer run at 90b1961 given this optimization.